### ISSUE TITLE
Document destruction of array

### ISSUE THREAD

**Issue by LinusU** [mafintosh/random-iterate#1]
Since the array is being modified I think that it would be good to document that.

Would you accept a pull request that changes the code to not modify the array?

**Comment by mafintosh** [comment]
can this be done without copying/iterating the entire array? i agree we should mention that it mutates the array :)

**Comment by LinusU** [comment]
We can store another list that contains the offsets of the selected items. Then select ran... let me show in code:

``` javascript
var taken = []

function next() {
  var max = list.length - taken.length
  var i = (Math.random() * max) | 0

  taken.forEach((offset) => (i <= offset) && (i++))
  taken.push(i)

  return list[i]
}
```

http://en.wikipedia.org/wiki/Fisher–Yates_shuffle

**Comment by sindresorhus** [comment]
Faster one https://github.com/sindresorhus/array-shuffle/blob/6be2549659220eee7d84a5810d0cd58f272954c8/index.js#L7-L19

**Comment by LinusU** [comment]
That copies the array (`var ret = arr.slice();`) which is not what @mafintosh asked for.

I'm guessing that the approach I posted would be more memory efficient, especially if using a typed array for `taken`.

**edit** It would also need to shuffle the entire array before the first item can be given, no?

**Comment by mafintosh** [comment]
@sindresorhus my usecase is iterating a possibly huge array to pick 5 random elements. i can do that now my doing:

```
var randomElements = []
var ite = randomIterator(hugeArray)
for (var i = 0; i < 5; i++) randomElements.push(ite())
```

@LinusU the algo you posted doesn't guarantee that it'll iterate through all elements in the list. fx if the first element it picks is the last one it won't find anything else

**Comment by LinusU** [comment]
@mafintosh That is not true, unless I made some implementation error. The [wikipedia article](https://en.wikipedia.org/wiki/Fisher–Yates_shuffle) describes it quite well.

> 1. Write down the numbers from 1 through N.
> 2. Pick a random number k between one and the number of unstruck numbers remaining (inclusive).
> 3. Counting from the low end, strike out the kth number not yet struck out, and write it down elsewhere.
> 4. Repeat from step 2 until all the numbers have been struck out.
> 5. The sequence of numbers written down in step 3 is now a random permutation of the original numbers.

We would pause at number 4 until the function is called the next time.

**Comment by jimmywarting** [comment]
Okey this was fun...

I figured I wanted a real iterator instead of a anonymous function instead so i could use `for..of`. 

IMO I that is better then returning an anonymous function all the time since the compiler won't be able to optimize a function that is recreated every time. A iterator can also be suspended and you can get the next value by calling `iterator.next()`

Then I looked at your code an saw that the array is mutated. So I found this issue and took all your considerations into account.

- [x] You didn't wanted to copy (slice) the array
- [x] You needed to store what values you have already taken
- [x] You wanted to use typed arrays to store the taken values
- [x] If the random value is already taken - get another value
- [x] i wanted to use `for..of` meaning it has to return a generator/iterator or having a Symbol.iterator

Using @LinusU example as a base referenser was not ideal since it logged the same value `e` 3 times when i tested it<details>
<summary>Result:</summary>

![skarmavbild 2018-09-07 kl 13 24 40](https://user-images.githubusercontent.com/1148376/45216743-e66a3900-b2a1-11e8-9e47-8947eb9f0695.png)

</details>
<br><br>

One conclusion was that I didn't just want to use a typed array whose length are the same as the array length `new Uint8Array(array.length)` i wanted to use all the available 8 bytes in an unsigned byte array. So if the array length was 10 - then it means that the byte array needs to be 2 to fit all the bits needed

I took some help from [bitfield](https://github.com/fb55/bitfield/)
<details>
<summary>So here is the final result</summary>

```js
function* randomIterator (arr) {
  var len = arr.length
  var out = len >> 3 // calc the smallest available size
  var buffer = new Uint8Array( (out % 8 !== 0) ? out+1:out )
  var i, j, L, R, offset, iterations = 0
  
  var get = i => buffer[j] & (128 >> (i % 8))
  var set = i => buffer[j] |= 128 >> (i % 8)

  while (iterations < len) {
    // Let i be the rolled dice
    i = (Math.random() * len) | 0

    // increase the offset of i and look
    // both to the left and right of the array
    // until it finds a item that is not taken
    for (offset = 0; true; offset++) {
      L = i - offset
      R = i + offset

      // look right
      if (R < len) {
        j = R >> 3
        
        if (!get(R)) {
          set(i = R)
          break
        }
      }
      
      // look left
      if (L >= 0) {
        j = L >> 3
        
        if (!get(L)) {
          set(i = L)
          break
        }
      }
    }

    iterations++
    yield arr[i]
  }
}
```

</details>

<details>
<summary>Then after running it throught Closure Compiler adv optimizations</summary>

```js
function* randomIterator(h) {
  var f = h.length,
    a = f >> 3;
  a = new Uint8Array(0 !== a % 8 ? a + 1 : a);
  for (var c, b, d, e, g, k = 0; k < f;) {
    c = Math.random() * f | 0;
    for (g = 0;; g++) {
      d = c - g;
      e = c + g;
      if (e < f && (b = e >> 3, !(a[b] & 128 >> e % 8))) {
        a[b] |= 128 >> (c = e) % 8;
        break
      }
      if (0 <= d && (b = d >> 3, !(a[b] & 128 >> d % 8))) {
        a[b] |= 128 >> (c = d) % 8;
        break
      }
    }
    k++;
    yield h[c]
  }
}
```

</details>

<details>
<summary>Then after it's minified</summary>

```js
function*randomIterator(h){var f=h.length,a=f>>3;a=new Uint8Array(0!==a%8?a+1:a);for(var c,b,d,e,g,k=0;k<f;){c=Math.random()*f|0;for(g=0;;g++){d=c-g;e=c+g;if(e<f&&(b=e>>3,!(a[b]&128>>e%8))){a[b]|=128>>(c=e)%8;break}if(0<=d&&(b=d>>3,!(a[b]&128>>d%8))){a[b]|=128>>(c=d)%8;break}}k++;yield h[c]}};
```

</details>

```js
// the usage is a little bit different:
arr = [1,2,3,4,5,6,7,8,9]
var ite = iterate(arr)

console.log(ite.next()) // maybe {value: 4, done: false}
console.log(ite.next()) // maybe {value: 9, done: false}
... 7 more time ...
console.log(ite.next()) // returns {value: undefined, done: true}
console.log(arr) // still the same

// or like this:
for (let random of iterate(arr)) {
  console.log(random)
}

// or
var shuffled  = [...iterate(arr)]
console.log(shuffled) // maybe [9,3,8,7,1,6,2,4,5]
```

if you want to keep the same backwards compatible with support for `for..of`

```js

var iterate = function (list) {
  var ite = randomIterator(list)
  function fn() {
    var res = ite.next()
    return res.done ? null : res.value
  }
  fn[Symbol.iterator] = function() {
    return ite
  }
  return fn
}
``` 

it could even be more friendlier if my `randomIterator` yielded a array with `[index, value]` instead and have it do the same thing as `Array.prototype.values` `Array.prototype.keys` and `Array.prototype.entries` and `Array.prototype[symbol.iterator]`

```js
var iterate = function (list) {
  var ite = randomIterator(list)
  function fn() {
    var res = ite.next()
    return res.done ? null : res.value
  }
  fn.values = function* () {
    for (let random of ite) yield random[1]
  }
  fn.keys = function* () {
    for (let random of ite) yield random[0]
  }
  fn.entries = function* () {
    for (let random of ite) yield random
  }
  fn[Symbol.iterator] = function() {
    return fn.value()
  }
  return fn
}
```
